[SW-2019] Implement auto cluster start for External Backend on Kubernetes & test

jakubhava opened the pull request.

mn-mikke commented:
cc @Pscheidl

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/external/KubernetesUtils.scala:
```diff
@@ -0,0 +1,309 @@
+  }
+
+  private def installH2OStateFulSet(client: KubernetesClient, conf: H2OConf): String = {
+    client
```
I prefer the programmatic way. To make it more readable I can split it to more smaller methods, but I would still keep it like this. Spark is using the same API as well

jakubhava replied:
I have split the code into more traits and split long definitions into smaller methods, I hope easier to read now.

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala:
```diff
@@ -90,6 +94,30 @@ trait ExternalBackendConf extends SharedBackendConf with Logging {
+  def externalK8sH2OApiPort: Int =
+    sparkConf.getInt(PROP_EXTERNAL_K8S_H2O_API_PORT._1, PROP_EXTERNAL_K8S_H2O_API_PORT._2)
+
+  def externalK8sNamespace: String = sparkConf.get(PROP_EXTERNAL_K8S_NAMESPACE._1, PROP_EXTERNAL_K8S_NAMESPACE._2)
```
Not required for the external backend, it can be in different namespace

mn-mikke replied:
If spark executors be in one namespace and h2o node in another one, Would they be able to communicate with each other.

jakubhava replied:
I will edit the test to run the external backend in different namespace then Spark.

As long as Spark is running also in K8s, I think we should be able to access H2O via the headless service DNS which is published. Will double check

jakubhava replied:
Updated the test so we have this automatically checked. Als o verified manually. The namespaces are really independent as we refer to the cluster via the service DNS.

jakubhava commented on ci/Jenkinsfile-kubernetes:
```diff
@@ -332,68 +349,142 @@ static String externalBackendSharedSubmitCmd(registryId, master, version, mode,
+def testPythonExternalBackendAutoClientMode(registryId, master, version) {
+    sh "kubectl delete pod sparkling-water-app --ignore-not-found"
+    installSparkHeadlessService()
+    def image = "${registryId}.dkr.ecr.us-east-2.amazonaws.com/sw_kubernetes_repo/sparkling-water:python-${version}"
```
our infra is running in us-east-2, so it will be always this region. I can, however, extract it into a variable

mn-mikke replied:
Is there any gradle property for that?

jakubhava replied:
Nope, but I will extract it into a variable in Jenkinsfile-Kubernetes and pass it to both terraform apply & the rest of the code that uses it so we avoid using the constant

jakubhava replied:
Done, thanks for the good suggestion!

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/external/KubernetesUtils.scala:
```diff
@@ -0,0 +1,309 @@
+
+  private def waitForClusterToBeReady(client: KubernetesClient, conf: H2OConf): String = {
+    while (!H2OStatefulSetExist(client, conf)) {
+      Thread.sleep(100)
```
I couldn't fine any

jakubhava replied:
Only a few resources have `waitUntilReady` method so I decided to keep it consistent and use the wait as it is now.

But I will edit the code pieces to have some ultimate timeout so it does not wait if something went wrong with for example service creation

jakubhava commented on assembly/build.gradle:
```diff
@@ -62,6 +62,7 @@ shadowJar {
   relocate 'scala.concurrent.java8', 'ai.h2o.scala.concurrent.java8'
   relocate 'com.amazonaws', 'ai.h2o.com.amazonaws'
   relocate 'org.apache.http', 'ai.h2o.org.apache.http'
+  relocate 'io.fabric8', 'ai.h2o.io.fabric8'
```
Relocating as there is a conflict with the same library used in Spark

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala:
```diff
@@ -172,6 +203,47 @@ trait ExternalBackendConf extends SharedBackendConf with Logging {
+    set(PROP_EXTERNAL_K8S_H2O_API_PORT._1, port.toString)
+  }
+
+  def externalK8sNamespace(namespace: String): H2OConf = {
```
Thanks!

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala:
```diff
@@ -172,6 +203,47 @@ trait ExternalBackendConf extends SharedBackendConf with Logging {
+    set(PROP_EXTERNAL_K8S_NAMESPACE._1, namespace)
+  }
+
+  def setExternalK8sDockerImage(name: String): H2OConf = {
```
By default, it points to the image we publish as part of our release pipeline, but user can suply custom image with the external backend

jakubhava commented:
@mn-mikke I still need to edit the tests to run in different namespaces and verify your questions, but apart from that, I tried to improve the code according to the suggestions where it was possible. Can you please have a second look?

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendConf.scala:
```diff
@@ -269,4 +351,40 @@ object ExternalBackendConf {
+  val PROP_EXTERNAL_K8S_EXPOSE_LEADER: (String, Boolean) = ("spark.ext.h2o.external.k8s.expose.leader", false)
+
+  /** Timeout in seconds used as a limit for K8s service creation */
+  val PROP_EXTERNAL_K8S_SERVICE_TIMEOUT: (String, Int) = ("spark.ext.h2o.external.k8s.svc.timeout", 60 * 5)
```
TODO for me: expose getter and seter on python and R side and in doc

Pscheidl commented:
We had a discussion about this with Kuba - H2O has no driver yet, so this is required. We'll soon have external XGBoost, which will simplify memory usage of H2O.

mn-mikke commented:
@Pscheidl Are you saying that the specification of H2O statefulset can't be reused from H2O codebase? https://github.com/h2oai/sparkling-water/pull/2200/files#diff-0535754e9f2dca7f75b04c34687d1494R30

jakubhava commented:
Kubernetes tests passed 🎉 . Please let me know if you have any other suggestions, otherwise, I will go ahead and merge

mn-mikke commented:
Apart from synchronous waiting and illegible K8s configuration, I don't have other comments.

jakubhava commented:
Regarding the waiting. I probably don't understand the motivation. We need to wait during the starting of the H2OStateful set synchronously, as we need to wait until the backend is ready. We do the same wait during starting Internal H2O backend and H2O does very similar wait in case of the driver.

 However, after your comment, I tried to remove the waits surrounding services etc. I removed the synchronous wait after deleting resources as the library already handles it. I can experiment with removing the wait after the service creation -> that might be possibly removed as well.

Does this make sense? 

Regarding the config -> 

The reason why I'm using the programmatic approach is that a lot of details in the resources are parametrized and at that point, it seemed better to me to have it in the code completely. The other approach would be to store the YAML template as resources and mark the parameters for substitution. WDYT @mn-mikke ?

jakubhava commented:
As I don't have really strong preferences, I will change the k8s configuration into YAML files. The library supports it :)

```
Service loadedSvc = client.services().load(getClass)(.getResourceAsStream("/path/to/file")).get()
```

mn-mikke commented:
> Regarding the waiting. I probably don't understand your motivation. We need to wait during the starting of the H2OStateful set synchronously, as we need to wait until the backend is ready. We do the same wait during starting of internal H2O backend and H2O does very similar wait in case of the driver.

This comment was rather about exploring whether the API you're using offers blocking or asynchronos methods. Just can't find an answer to my original question.

> The reason why I'm using the programmatic approach is that a lot of details in the resources are parametrized and at that point, it seemed better to me to have it in the code completely. 

The programatic configuration is very difficult to understande to other people. Even @Pscheidl, who is the author of the original configuration, had troubles to understand what the configuration describes. IMHO, YAML or JSON would be definitely more convenient for other people.

jakubhava commented:
Thanks for the reply!

Add 1) The api doesn't provide that
Add 2) I don't see any comment from @Pscheidl here, but I guess there was some different communication I wasn't part of. Will change to YAML format...

Pscheidl commented:
Just a quick note: Make sure to you parallel pod management policy for stateful sets. It spawns the pods MUCH faster.

jakubhava commented:
Thank you @pavel on the Parallel hint! Done :)

@mn-mike can you please have another look? I have removed all waits except one during creation of statefulset. That wait is used for waiting for the cluster.

I have also removed the programmatic configuration in favor of a more readable solution. Thanks for pushing for this, I have to agree this is much better :)
